java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java

```diff
@@ -385,10 +392,97 @@ private API getXPAPI(
+        || isSymbolAndMatchesFlagName(argSym)
+        || isMatchingEnumFieldValue(argTree, state));
+  }
+
+  /*
+   * Matches on an enum in an expression tree. Covers the following 4 cases:
+   * 1. Unqualified import enum constant (e.g. "STALE_FLAG")
+   * 2. Qualified enum name (e.g. "TestExperimentName.STALE_FLAG")
+   * 3. Unqualified enum constant with method call (e.g. "STALE_FLAG.getKey()")
+   * 4. Qualified enum name with method call (e.g. "TestExperimentName.STALE_FLAG.getKey()")
+   */
+  private boolean isMatchingEnumFieldValue(ExpressionTree argTree, VisitorState state) {
+    // Skip if no enum records are configured, to avoid excessive AST lookups
+    if (!this.config.hasEnumRecords()) {
+      return false;
+    }
+
+    Symbol sym;
+    if (argTree.getKind() == Kind.IDENTIFIER) {
+      // Case 1: Unqualified enum constant (e.g. "STALE_FLAG")
+      sym = ASTHelpers.getSymbol(argTree);
+    } else if (argTree.getKind() == Kind.MEMBER_SELECT) {
+      // Case 2: Qualified enum name (e.g. "TestExperimentName.STALE_FLAG")
+      MemberSelectTree memberTree = (MemberSelectTree) argTree;
+      sym = ASTHelpers.getSymbol(memberTree);
+    } else if (argTree.getKind() == Kind.METHOD_INVOCATION) {
+      MethodInvocationTree methodTree = (MethodInvocationTree) argTree;
+      if (methodTree.getMethodSelect().getKind() != Kind.MEMBER_SELECT) {
+        return false;
+      }
+      MemberSelectTree memberTree = (MemberSelectTree) methodTree.getMethodSelect();
+      if (memberTree.getExpression().getKind() == Kind.IDENTIFIER) {
+        // Case 3: Unqualified enum constant with method call (e.g. "STALE_FLAG.getKey()")
+        IdentifierTree identifier = (IdentifierTree) memberTree.getExpression();
+        sym = ASTHelpers.getSymbol(identifier);
+      } else if (memberTree.getExpression().getKind() == Kind.MEMBER_SELECT) {
+        // Case 4: Qualified enum name with method call (e.g.
+        // "TestExperimentName.STALE_FLAG.getKey()")
+        MemberSelectTree memberTreeExpression = (MemberSelectTree) memberTree.getExpression();
+        sym = ASTHelpers.getSymbol(memberTreeExpression);
+      } else {
+        return false;
+      }
+    } else {
+      return false;
+    }
+
+    if (sym.getKind() != ElementKind.ENUM_CONSTANT) {
+      return false;
+    }
+
+    Symbol.ClassSymbol enclosingClass = ASTHelpers.enclosingClass(sym);
+
+    ImmutableCollection<PiranhaEnumRecord> enumRecords =
+        this.config.getEnumRecordsForName(enclosingClass.getSimpleName().toString());
+
+    if (!enumRecords.isEmpty()) {
+      return isEnumSymbol(sym.getSimpleName().toString(), enclosingClass, state, enumRecords);
+    }
+
+    return false;
+  }
+
+  public boolean isEnumSymbol(
+      String enumName,
+      Symbol.ClassSymbol classSymbol,
+      VisitorState state,
+      ImmutableCollection<PiranhaEnumRecord> enumRecords) {
+    ClassTree enumClassTree = ASTHelpers.findClass(classSymbol, state);
+    if (enumClassTree == null) {
```

Comment:
So far I tried to query members of the symbol for the enum class and was able to get the members. However, it seems that the symbol table doesn't hold the actual `"stale.flag"` String value from the constructor.  At least I don't see it in whatever implementation of the JVM I'm on (a version of AdoptOpenJDK 1.8) in the `com.sun.tools.javac.code` package. But I assume the String value has to be accessible outside of source (such as in the bytecode), so I'll keep investigating on how to get it for other targets/jars.

Reply:
Kind of thinking about this later, we need the source to be able to remove the enum constant in the first place? Does Piranha already have tests that cover the case where we have an enum in another target/jar, and therefore possibly no access to source code, but we clean up the enum's usages anyway while preserving the enum constant?

I'm really not sure how to get access to the constructor argument values from the symbol table. I don't see it represented anywhere in memory.

To avoid having this block the rest of the PR, may be another ticket to do later, and I can add a TODO if that's OK? At least it will work for projects that have their flags self-contained. I think regardless it would be good to add tests for feature flags across multi-module projects like your example.